### PR TITLE
[FIX] Enable CORS for Restivus

### LINKED ISSUE #7915: CORS broken: OPTIONS requests returns 500 Internal Server Error
Rocket.Chat version: 0.58.0

Anyway, I am trying to make an API call from the browser. It sends an OPTIONS request before POST to /api/logon to check CORS. The thing is, I am getting 500 Internal Server Error on this OPTIONS request, which is rendering Rocket Chat impossible to access from the browser.

### PR DESCRIPTION
**Pull request by mrsimpson**
@RocketChat/core 

Closes #7915

Previously, the setting to enable CORS on the API (`Admin > General > REST Api`) added the CORS-header, but the information was not propagated to Restivus.

### CODE REVIEW

**Comment by mrsimpson** [comment]
@graywolf336 I'm not really knowledgable about CORS and what the option in Restivus is.
Thus, I was confused by the setting not being propagated like I did now earlier - should the enabling in Restivus be a second setting (semantically)?

**Comment by graywolf336** [comment]
@mrsimpson only issue with this is that if you change it after the server has started, it won't be reflected until you restart the server. :thinking: so we need to figure out how to dynamically have this setting applied, I will take a look.

**Comment by mrsimpson** [comment]
@graywolf336 ah, Nö server side reactivity... how about getting the setting to a variable and set this variable additionally in a setting.load()-callback?

File: packages/rocketchat-api/server/api.js [hunk 1]
```diff
@@ -178,17 +178,26 @@ const getUserAuth = function _getUserAuth() {
...
+};
 
-RocketChat.API.default = new API({
-	useDefaultAuth: true,
-	prettyJson: true,
-	enableCors: false,
-	auth: getUserAuth()
+RocketChat.settings.onload('API_Enable_CORS', (key, value) => {
+	enableCors = value;
+	createApi();
```
Comment (mrsimpson) [inline packages/rocketchat-api/server/api.js]: @graywolf336 that's what I meant writing y'days comment. 
Is that the proper way to do it?

_Note to myself: I really should get an IDE on my mobile..._

Comment (mrsimpson) [inline packages/rocketchat-api/server/api.js]: @graywolf336 that's what I meant writing y'days comment. 
Is that the proper way to do it?

_Note to myself: I really should get an IDE on my mobile..._

Comment (graywolf336) [reply packages/rocketchat-api/server/api.js]: You're extremely close! Honestly, I would leave the function the way it is and just change the `RocketChat.settings.onload` handler to be `RocketChat.settings.get('API_Enabled_CORS', () => {})` as that will be called everytime the setting is changed. I have no clue what `onload` will do :thinking:

Comment (graywolf336) [reply packages/rocketchat-api/server/api.js]: So, after doing some review it seems that `get` with a callback makes a call to `onload` and also immediately runs the callback..someone feel free to correct me if I'm misunderstanding things

Comment (mrsimpson) [reply packages/rocketchat-api/server/api.js]: Is get() really being triggered after update of the setting? onLoad() surely is

Comment (graywolf336) [reply packages/rocketchat-api/server/api.js]: @rodrigok which one is the preferred way to do what is being discussed here?

Comment (rodrigok) [reply packages/rocketchat-api/server/api.js]: @graywolf336 Yes, `get` with `callback` will be called everytime the setting changes

**Comment by rodrigok** [comment]
Did you guys tested what happens if you create the same API twice?

**Comment by mrsimpson** [comment]
@rodrigok you mean whther we tested that changing the CORS-Setting creates a working api?

**Comment by rodrigok** [comment]
@mrsimpson Yes, now the same API could be create multiple times, is that working ok? Restivus replace the old v1 namespace by the new one correctly?

**Comment by mrsimpson** [comment]
@rodrigok I logged in via API, changed the setting and logged out via API again. 
This was working fine. 
However, when looking at the headers themselves, I noticed that the „access-control-allow-origin with * as value was always present.

**Comment by rodrigok** [comment]
The CI is showing a problem that could help by the delayed API creation:
```
Exception in callback of async function: TypeError: Cannot read property 'helperMethods' of undefined
    at requestParams.js (packages/rocketchat:api/server/v1/helpers/requestParams.js:1:19)
    at fileEvaluate (packages/modules-runtime.js:333:9)
    at require (packages/modules-runtime.js:228:16)
    at /tmp/build-test/bundle/programs/server/packages/rocketchat_api.js:3962:1
```

Any ideas?

**Comment by mrsimpson** [comment]
@rodrigok I can't really understand this, to be honest.
The very same script (`.scripts/start.js`) is executed twice in CircleCI. Once it failed (without oplog), once it was successful (with oplog). I cannot imagine this being a true positive. 
Executed locally, the tests also succeed.

However, I noted that the delayed instantiation was done onLoad(), you recommended get(callback) some time back, so I changed that. let's see what the CI says now.

**Comment by rodrigok** [comment]
@mrsimpson That's why I said `could` happen. If you are delaying the registration of one API like `RocketChat.API.v1`, other part of the code could try to use if before the registration.

The solution for that, IMO, is register the normal API as before, without CORS enabled and then use the settings callback to register it again if different from previous registration, in this first case if the setting is true cuz the default registration was false. You could use a control variable to do that or check if you can access the config like `RocketChat.API.v1.cors === true`.

**Comment by mrsimpson** [comment]
@rodrigok you seem to have been right with the failure being caused by the delayed API creation:
I added it outside the callback as well and the CI is happy now. 
This has the effect of the API being created twice on startup, but since this has only minimal impact on performance and no functional limitations, I believe it should be fine.

**Comment by rodrigok** [comment]
Oh, I just read your changes, that is almost what I said, can you change to only create the api if the setting value is different of the current value? And check if the first time you get the value from the setting in a sync way if the value is not undefined (set to false if undefined)?

**Comment by mrsimpson** [comment]
Amazing, how much ~bugs can be killed~ one can invest into five lines of code...
Updated, tested, pushed.

I don't like accessing a "private" property for deciding about the buffering, particularly in an untyped world, but I preferred this to buffering the property in some other place.

File: packages/rocketchat-api/server/api.js [hunk 2]
```diff
@@ -180,17 +180,31 @@ const getUserAuth = function _getUserAuth() {
...
 
-RocketChat.API.v1 = new API({
-	version: 'v1',
-	useDefaultAuth: true,
-	prettyJson: true,
-	enableCors: false,
-	auth: getUserAuth()
-});
+const createApi = function(enableCors) {
+	if (!RocketChat.API.v1 || RocketChat.API.v1._config.enableCors !== enableCors) {
```
Comment (mrsimpson) [inline packages/rocketchat-api/server/api.js]: I don't like accessing a "private" property for deciding about the buffering, particularly in an untyped world, but I preferred this to buffering the property in some other place.

**Comment by rodrigok** [comment]
Thanks @mrsimpson It looks good 😄

**Pull request merged**